# Import librerias

In [1]:
import requests
import json
import pandas as pd
from collections import Counter
import numpy as np
import seaborn as sns
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
from datetime import datetime

In [2]:
print("Trabajaremos con las siguientes versiones\n")
print("Pandas",pd.__version__)
print("Numpy",np.__version__)
print("Seaborn", sns.__version__)
!python --version

Trabajaremos con las siguientes versiones

Pandas 1.0.5
Numpy 1.18.5
Seaborn 0.10.1
Python 3.8.3


# Funciones

In [3]:
def import_data(resource_id):

    #Define base url and parameters
    url = 'https://opendata-ajuntament.barcelona.cat/data/api/action/datastore_search'
    params = {'resource_id': resource_id, 'limit': 100000000000000} # Change limit size of the retrieved rows: https://docs.ckan.org/en/latest/maintaining/datastore.html#the-data-api  (see parameters in the datastore_search request)

    #Conect to API
    response_API = requests.get(url, params=params)
    if response_API.status_code != 200:
        raise Exception('Could not get data from the API')

    #Get data from API
    data = response_API.json()

    #Get data records
    data_records = data['result']['records']

    #Convert json into a dataframe
    df = pd.json_normalize(data_records)

    return df

In [4]:
# Creamos una función simple para hacer un report simple sobre los DataFrame suministrados
def report_df(dataframe, verbose = True):
    print("El tamaño del dataset es", dataframe.shape[0], "filas y", dataframe.shape[1], "columnas")
    print()
    print(dataframe.info(verbose = verbose))
    total_nulos = dataframe.isnull().sum().sum()
    print()
    print(f"Tenemos un total de {total_nulos} nulos")
    print()
    print("Hay", dataframe.duplicated().sum(), "duplicados")


# CensComercialBCN 2014

## Load data

In [5]:
df_cens = import_data('c933fa10-a965-4509-8de6-e519281dc245').drop(columns = '_id')
df_cens.head()

Y_ED50              NOM CBARRI  \
0  4584527,7           Orange     69   
1  4584524,4       Local buit     69   
2    4584520  La Llave de oro     69   
3  4584515,5       Local buit     69   
4  4584511,1       Local buit     69   

                                         NBARRI   CARRCADAST CODI_EIXOS  \
0  Diagonal Mar i el Front Marítim del Poblenou  AV DIAGONAL       movi   
1  Diagonal Mar i el Front Marítim del Poblenou  AV DIAGONAL     localb   
2  Diagonal Mar i el Front Marítim del Poblenou  AV DIAGONAL       inmo   
3  Diagonal Mar i el Front Marítim del Poblenou  AV DIAGONAL     localb   
4  Diagonal Mar i el Front Marítim del Poblenou  AV DIAGONAL     localb   

         Y        RCNUMBER CDISTRI                      BCN3        BCN2  \
0  4584527  3644423DF3834D      10                    ALTRES      ALTRES   
1  4584524  3644423DF3834D      10                LOCAL BUIT  LOCAL BUIT   
2  4584520  3644426DF3834D      10  ACTIVITATS IMMOBILIÀRIES     SERVEIS   
3  4584515  3644426DF3834D      10                LOCAL BUIT  LOCAL BUIT   
4  4584511  3644426DF3834D      10                LOCAL BUIT  LOCAL BUIT   

         BCN1 DOORNUM       X ULTIMA_VISITA   DATAINICI    NDISTRIC    X_ED50  \
0     SERVEIS      89  433707    31/03/2014  2012-05-16  Sant Martí  433707,7   
1  LOCAL BUIT      89  433701    31/03/2014  2012-05-16  Sant Martí  433701,1   
2     SERVEIS      91  433678    31/03/2014  2012-05-16  Sant Martí  433678,8   
3  LOCAL BUIT      91  433664    31/03/2014  2012-05-16  Sant Martí  433664,4   
4  LOCAL BUIT      93  433649    31/03/2014  2012-05-16  Sant Martí  433649,9   

  M_CODI_EIXOS  
0         serv  
1         locb  
2         serv  
3         locb  
4         locb

## EDA y preprocesamiento

In [6]:
## Observamos los datos que tenemos 
report_df(df_cens)

df_cens.head(2)

El tamaño del dataset es 67433 filas y 19 columnas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67433 entries, 0 to 67432
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Y_ED50         67433 non-null  object
 1   NOM            67433 non-null  object
 2   CBARRI         67433 non-null  object
 3   NBARRI         67433 non-null  object
 4   CARRCADAST     67433 non-null  object
 5   CODI_EIXOS     67433 non-null  object
 6   Y              67433 non-null  object
 7   RCNUMBER       67433 non-null  object
 8   CDISTRI        67433 non-null  object
 9   BCN3           67433 non-null  object
 10  BCN2           67433 non-null  object
 11  BCN1           67433 non-null  object
 12  DOORNUM        67433 non-null  object
 13  X              67433 non-null  object
 14  ULTIMA_VISITA  67433 non-null  object
 15  DATAINICI      67433 non-null  object
 16  NDISTRIC       67433 non-null  object
 17  X_ED50         67

Y_ED50         NOM CBARRI                                        NBARRI  \
0  4584527,7      Orange     69  Diagonal Mar i el Front Marítim del Poblenou   
1  4584524,4  Local buit     69  Diagonal Mar i el Front Marítim del Poblenou   

    CARRCADAST CODI_EIXOS        Y        RCNUMBER CDISTRI        BCN3  \
0  AV DIAGONAL       movi  4584527  3644423DF3834D      10      ALTRES   
1  AV DIAGONAL     localb  4584524  3644423DF3834D      10  LOCAL BUIT   

         BCN2        BCN1 DOORNUM       X ULTIMA_VISITA   DATAINICI  \
0      ALTRES     SERVEIS      89  433707    31/03/2014  2012-05-16   
1  LOCAL BUIT  LOCAL BUIT      89  433701    31/03/2014  2012-05-16   

     NDISTRIC    X_ED50 M_CODI_EIXOS  
0  Sant Martí  433707,7         serv  
1  Sant Martí  433701,1         locb

In [7]:
# Examinamos los duplicados existentes
df_cens[df_cens.duplicated()]

Y_ED50                 NOM CBARRI  \
125      4583986,6          Local Buit     06   
150      4583601,1          Local Buit     07   
179        4583750          Local Buit     06   
208    4583706,654          Local Buit     07   
266      4583499,9          Local Buit     07   
...            ...                 ...    ...   
67374    4588284,4           Varietats     48   
67387  4582415,511  LLI privada Ninets     08   
67393    4587906,6           Varietats     48   
67400      4587970           Varietats     48   
67423  4585471,138   Castro de la Peña     35   

                                NBARRI       CARRCADAST CODI_EIXOS  \
125                 la Sagrada Família      AV DIAGONAL     localb   
150             la Dreta de l Eixample      AV DIAGONAL     localb   
179                 la Sagrada Família      AV DIAGONAL     localb   
208             la Dreta de l Eixample      AV DIAGONAL     localb   
266             la Dreta de l Eixample      AV DIAGONAL     localb   
...                                ...              ...        ...   
67374                     la Guineueta       RB CAÃADOR     variet   
67387  l Antiga Esquerra de l Eixample         ValÞncia       educ   
67393                     la Guineueta  RB FABRA I PUIG     variet   
67400                     la Guineueta  RB FABRA I PUIG     variet   
67423                      el Guinardó           Volart       educ   

                    Y        RCNUMBER CDISTRI         BCN3        BCN2  \
125           4583986  1740810DF3814B      02   LOCAL BUIT  LOCAL BUIT   
150           4583601  0835306DF3803F      02   LOCAL BUIT  LOCAL BUIT   
179        4583750.37  1037609DF3813G      02   LOCAL BUIT  LOCAL BUIT   
208       45837065477  0837728DF3803F      02   LOCAL BUIT  LOCAL BUIT   
266        4583499.89  0235724DF3803E      02   LOCAL BUIT  LOCAL BUIT   
...               ...             ...     ...          ...         ...   
67374         4588284  0682301DF3808D      08       ALTRES      ALTRES   
67387  45824151104439                      02  ENSENYAMENT     SERVEIS   
67393         4587906  0479903DF3807G      08       ALTRES      ALTRES   
67400         4587970  0479902DF3807G      08       ALTRES      ALTRES   
67423  45854713800091                      07  ENSENYAMENT     SERVEIS   

             BCN1  DOORNUM               X ULTIMA_VISITA   DATAINICI  \
125    LOCAL BUIT      298          431770    31/03/2014  2013-12-09   
150    LOCAL BUIT      319          430826    31/03/2014  2013-12-31   
179    LOCAL BUIT      336       431055.96    31/03/2014  2011-07-06   
208    LOCAL BUIT      358  43086864201411    31/03/2014  2013-12-31   
266    LOCAL BUIT      414       430274.69    31/03/2014  2011-07-06   
...           ...      ...             ...           ...         ...   
67374      ALTRES        1          430536    09/03/2014  2014-03-09   
67387     SERVEIS  218-220  42971322259258    16/04/2013  2013-04-16   
67393      ALTRES      474          430454    03/03/2014  2014-03-09   
67400      ALTRES      478          430463    03/03/2014  2014-03-09   
67423     SERVEIS       81  43109261141759    16/04/2013  2013-04-16   

             NDISTRIC      X_ED50 M_CODI_EIXOS  
125          Eixample      431770         locb  
150          Eixample    430826,6         locb  
179          Eixample    431055,5         locb  
208          Eixample  430868,864         locb  
266          Eixample    430274,4         locb  
...               ...         ...          ...  
67374      Nou Barris    430536,6       ALTRES  
67387        Eixample  429713,322        other  
67393      Nou Barris    430454,4       ALTRES  
67400      Nou Barris    430463,3       ALTRES  
67423  Horta-Guinardó  431092,261        other  

[989 rows x 19 columns]

In [8]:
# Comprobamos si estos duplicados, realmente lo son
df_cens[df_cens['RCNUMBER'] == '0837728DF3803F']

Y_ED50         NOM CBARRI                  NBARRI   CARRCADAST  \
207  4583706,654  Local Buit     07  la Dreta de l Eixample  AV DIAGONAL   
208  4583706,654  Local Buit     07  la Dreta de l Eixample  AV DIAGONAL   

    CODI_EIXOS            Y        RCNUMBER CDISTRI        BCN3        BCN2  \
207     localb  45837065477  0837728DF3803F      02  LOCAL BUIT  LOCAL BUIT   
208     localb  45837065477  0837728DF3803F      02  LOCAL BUIT  LOCAL BUIT   

           BCN1 DOORNUM               X ULTIMA_VISITA   DATAINICI  NDISTRIC  \
207  LOCAL BUIT     358  43086864201411    31/03/2014  2013-12-31  Eixample   
208  LOCAL BUIT     358  43086864201411    31/03/2014  2013-12-31  Eixample   

         X_ED50 M_CODI_EIXOS  
207  430868,864         locb  
208  430868,864         locb

In [9]:
# Decidimos eliminar duplicados, viendo que tiene sentido
df_cens.drop_duplicates(inplace=True)

In [10]:
# Eliminamos columnas que no hacen falta
df_cens =  df_cens.drop(['Y_ED50', 'X_ED50', 'X', 'Y', 
                         'CARRCADAST', 'DOORNUM',
                         'ULTIMA_VISITA', 'DATAINICI',
                         'RCNUMBER'], axis=1)

In [11]:
# Comprobamos los datos que contenemos 
df_cens.describe(exclude=np.number).T

count unique                     top   freq
NOM           66444  45556              Local Buit   9228
CBARRI        66444     74                      07   4142
NBARRI        66444     74  la Dreta de l Eixample   4142
CODI_EIXOS    66444    120                  localb   9715
CDISTRI       66444     11                      02  14931
BCN3          66444     41                  ALTRES  17930
BCN2          66444      9                 SERVEIS  30785
BCN1          66444      4                 SERVEIS  31244
NDISTRIC      66444     11                Eixample  14931
M_CODI_EIXOS  66444     11                    serv  16895

In [12]:
# Probamos los valores que contiene esta columna para entender mejor nuestro dataset
df_cens['NBARRI'].value_counts()

la Dreta de l Eixample             4142
la Vila de Gràcia                  3481
Sant Gervasi - Galvany             2947
l Antiga Esquerra de l Eixample    2810
el Raval                           2540
                                   ... 
#N/A                                 21
Baró de Viver                        21
Torre Baró                           20
Vallbona                              8
la Clota                              6
Name: NBARRI, Length: 74, dtype: int64

In [13]:
# Remplazamos los valores nulos para que se puedan detectar
df_cens.replace('#N/A',np.nan,inplace=True)

In [14]:
# Comprobamos que se hizo el cambio correctamente
df_cens[df_cens.isnull().any(axis=1)].head(2)

NOM CBARRI NBARRI CODI_EIXOS CDISTRI  \
3556   CLUB DE PETANCA CASA OLIVA    NaN    NaN       espo     NaN   
14999        Institut Lluís Vives    NaN    NaN       educ     NaN   

              BCN3     BCN2     BCN1 NDISTRIC M_CODI_EIXOS  
3556        ALTRES   ALTRES   ALTRES      NaN       ALTRES  
14999  ENSENYAMENT  SERVEIS  SERVEIS      NaN        other

In [15]:
# Eliminamos valores nulos
df_cens.dropna(inplace=True)
df_cens.reset_index(drop=True,inplace=True)

In [16]:
# Renombramos columnas para trabajar más fácilmente
df_cens.rename(columns={'CBARRI': 'Codi_Barri', 'NBARRI': 'Nom_Barri', 'CDISTRI': 'Codi_Districte', 'NDISTRIC': 'Nom_Districte'}, inplace=True)

In [17]:
df_cens.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66423 entries, 0 to 66422
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   NOM             66423 non-null  object
 1   Codi_Barri      66423 non-null  object
 2   Nom_Barri       66423 non-null  object
 3   CODI_EIXOS      66423 non-null  object
 4   Codi_Districte  66423 non-null  object
 5   BCN3            66423 non-null  object
 6   BCN2            66423 non-null  object
 7   BCN1            66423 non-null  object
 8   Nom_Districte   66423 non-null  object
 9   M_CODI_EIXOS    66423 non-null  object
dtypes: object(10)
memory usage: 5.1+ MB


In [18]:
# Comprobamos diferencias entre columnas BCN3, BCN2 y BCN1
df_cens[(df_cens['NOM']=='Local Buit') & (df_cens['BCN2']!= 'LOCAL BUIT')]

NOM Codi_Barri                      Nom_Barri CODI_EIXOS  \
2912   Local Buit         27           el Putxet i el Farró       moda   
3833   Local Buit         19                      les Corts       ofic   
9869   Local Buit         25     Sant Gervasi - la Bonanova       moda   
11022  Local Buit         26         Sant Gervasi - Galvany       moda   
12407  Local Buit         17                  Sants - Badal      colma   
12413  Local Buit         46            el Turó de la Peira       locu   
15234  Local Buit         60                    Sant Andreu       tall   
16017  Local Buit         18                          Sants       inmo   
18470  Local Buit         44  Vilapicina i la Torre Llobeta       ofic   
19589  Local Buit         18                          Sants     arranj   
21663  Local Buit         44  Vilapicina i la Torre Llobeta       ofic   
24725  Local Buit         27           el Putxet i el Farró       info   
26277  Local Buit         65                        el Clot     constr   
27305  Local Buit         51                         Verdun     constr   
29041  Local Buit         01                       el Raval       pelu   
30094  Local Buit         37                      el Carmel     constr   
32693  Local Buit         01                       el Raval     constr   
33221  Local Buit         18                          Sants     constr   
33600  Local Buit         18                          Sants       conc   
43532  Local Buit         18                          Sants       tall   
47028  Local Buit         44  Vilapicina i la Torre Llobeta       ofic   
47396  Local Buit         18                          Sants       cong   
54797  Local Buit         27           el Putxet i el Farró       inmo   
58796  Local Buit         71        Provençals del Poblenou       naub   
58799  Local Buit         71        Provençals del Poblenou       naub   
58841  Local Buit         01                       el Raval     audiov   
64543  Local Buit         73            la Verneda i la Pau       locu   

      Codi_Districte                                         BCN3  \
2912              05                                       VESTIR   
3833              04                                       ALTRES   
9869              05                                       VESTIR   
11022             05                                       VESTIR   
12407             03                                  ALIMENTACIÓ   
12413             08                                       ALTRES   
15234             09  REPARACIONS (ELECTRODOMÈSTICS I AUTOMÒBILS)   
16017             03                     ACTIVITATS IMMOBILIÀRIES   
18470             08                                       ALTRES   
19589             03              MANTENIMENT, NETEJA I PRODUCCIÓ   
21663             08                                       ALTRES   
24725             05                                  INFORMÀTICA   
26277             10                                       ALTRES   
27305             08                                       ALTRES   
29041             01                                       ALTRES   
30094             07                                       ALTRES   
32693             01                                       ALTRES   
33221             03                                       ALTRES   
33600             03                                     VEHICLES   
43532             03  REPARACIONS (ELECTRODOMÈSTICS I AUTOMÒBILS)   
47028             08                                       ALTRES   
47396             03                                  ALIMENTACIÓ   
54797             05                     ACTIVITATS IMMOBILIÀRIES   
58796             10                                       ALTRES   
58799             10                                       ALTRES   
58841             01                                      MÚSICA*   
64543             10                                       ALTRES   

        

In [19]:
# Entender mejor los datos que contiene, vemos que se parece a las columnas de BCN3,2 y 1
df_cens['M_CODI_EIXOS'].value_counts()

serv      16895
host      10249
locb       9715
pers       6835
quot       6104
ALTRES     4069
altr       2958
llar       2785
tran       2522
cult       2499
other      1792
Name: M_CODI_EIXOS, dtype: int64

In [20]:
df_cens['BCN1'].value_counts()

SERVEIS             31239
COMERÇ AL DETALL    21400
LOCAL BUIT           9715
ALTRES               4069
Name: BCN1, dtype: int64

In [21]:
df_cens['BCN3'].value_counts()

ALTRES                                                            17914
RESTAURANTS, BARS I HOTELS (INCLÒS HOSTALS, PENSIONS I FONDES)    10249
LOCAL BUIT                                                         9715
VESTIR                                                             3488
ALIMENTACIÓ                                                        2406
ENSENYAMENT                                                        2239
PA, PASTISSERIA I LÀCTICS                                          1705
MOBLES I ARTICLES FUSTA I METALL                                   1639
FINANCES I ASSEGURANCES                                            1551
SANITAT I ASSISTÈNCIA                                              1385
REPARACIONS (ELECTRODOMÈSTICS I AUTOMÒBILS)                        1283
LLIBRES, DIARIS I REVISTES                                         1125
FARMÀCIES                                                          1065
FRUITES I VERDURES                                              

In [22]:
df_cens['BCN2'].value_counts()

SERVEIS                   30780
LOCAL BUIT                 9715
QUOTIDIÀ ALIMENTARI        6569
ALTRES                     6277
EQUIPAMENT PERSONAL        5114
PARAMENT DE LA LLAR        3086
OCI I CULTURA              2798
QUOTIDIÀ NO ALIMENTARI     1724
AUTOMOCIÓ                   360
Name: BCN2, dtype: int64

In [23]:
df_cens.columns

Index(['NOM', 'Codi_Barri', 'Nom_Barri', 'CODI_EIXOS', 'Codi_Districte',
       'BCN3', 'BCN2', 'BCN1', 'Nom_Districte', 'M_CODI_EIXOS'],
      dtype='object')

In [24]:
# Eliminamos más columnas que vimos que contiene valores parecidos para reducir nuestro dataset de trabajo
df_cens.drop(['NOM','CODI_EIXOS','BCN3','BCN1','M_CODI_EIXOS'],axis=1,inplace=True)

In [25]:
df_cens.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66423 entries, 0 to 66422
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Codi_Barri      66423 non-null  object
 1   Nom_Barri       66423 non-null  object
 2   Codi_Districte  66423 non-null  object
 3   BCN2            66423 non-null  object
 4   Nom_Districte   66423 non-null  object
dtypes: object(5)
memory usage: 2.5+ MB


In [26]:
# Cambio nombre columna por un nombre más adecuado al contenido
df_cens.rename(columns = {'BCN2':'Activitat_Economica'},inplace=True)

In [27]:
df_cens.head()

Codi_Barri                                     Nom_Barri Codi_Districte  \
0         69  Diagonal Mar i el Front Marítim del Poblenou             10   
1         69  Diagonal Mar i el Front Marítim del Poblenou             10   
2         69  Diagonal Mar i el Front Marítim del Poblenou             10   
3         69  Diagonal Mar i el Front Marítim del Poblenou             10   
4         69  Diagonal Mar i el Front Marítim del Poblenou             10   

  Activitat_Economica Nom_Districte  
0              ALTRES    Sant Martí  
1          LOCAL BUIT    Sant Martí  
2             SERVEIS    Sant Martí  
3          LOCAL BUIT    Sant Martí  
4          LOCAL BUIT    Sant Martí

In [28]:
report_df(df_cens)

El tamaño del dataset es 66423 filas y 5 columnas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66423 entries, 0 to 66422
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Codi_Barri           66423 non-null  object
 1   Nom_Barri            66423 non-null  object
 2   Codi_Districte       66423 non-null  object
 3   Activitat_Economica  66423 non-null  object
 4   Nom_Districte        66423 non-null  object
dtypes: object(5)
memory usage: 2.5+ MB
None

Tenemos un total de 0 nulos

Hay 65818 duplicados


## Cálculo de actividad por barrio

In [29]:
# Groupby para calcular el total de actividades que hay según el barrio y el tipo de actividad economica
Activitat_Barris = df_cens.groupby(['Codi_Barri','Activitat_Economica']).size().to_frame(name = 'Total_Activitat_Per_Barri').reset_index()

In [30]:
# Imprimimos el dataframe generado para ver como queda y ver si tiene  sentido
Activitat_Barris

Codi_Barri     Activitat_Economica  Total_Activitat_Per_Barri
0           01                  ALTRES                        381
1           01               AUTOMOCIÓ                          1
2           01     EQUIPAMENT PERSONAL                        225
3           01              LOCAL BUIT                        245
4           01           OCI I CULTURA                         88
..         ...                     ...                        ...
600         73           OCI I CULTURA                         26
601         73     PARAMENT DE LA LLAR                         43
602         73     QUOTIDIÀ ALIMENTARI                        116
603         73  QUOTIDIÀ NO ALIMENTARI                         28
604         73                 SERVEIS                        356

[605 rows x 3 columns]

In [31]:
# Comprobación visual de un barrio aleatorio
Activitat_Barris[Activitat_Barris['Codi_Barri']=='69']

Codi_Barri     Activitat_Economica  Total_Activitat_Per_Barri
561         69                  ALTRES                         20
562         69     EQUIPAMENT PERSONAL                          8
563         69              LOCAL BUIT                         42
564         69           OCI I CULTURA                         17
565         69     PARAMENT DE LA LLAR                          3
566         69     QUOTIDIÀ ALIMENTARI                         19
567         69  QUOTIDIÀ NO ALIMENTARI                         11
568         69                 SERVEIS                        185

In [32]:
# merge del dataframe original con el dataframe generado con el groupby anterior. Comprobamos que no se haga un cartesiano
print(df_cens.shape)

df_cens = pd.merge(df_cens,Activitat_Barris, on = ['Codi_Barri','Activitat_Economica'], how = 'left')

print(df_cens.shape)

(66423, 5)
(66423, 6)


In [33]:
# imprimimos por pantalla para comprobar como queda
df_cens.head()

Codi_Barri                                     Nom_Barri Codi_Districte  \
0         69  Diagonal Mar i el Front Marítim del Poblenou             10   
1         69  Diagonal Mar i el Front Marítim del Poblenou             10   
2         69  Diagonal Mar i el Front Marítim del Poblenou             10   
3         69  Diagonal Mar i el Front Marítim del Poblenou             10   
4         69  Diagonal Mar i el Front Marítim del Poblenou             10   

  Activitat_Economica Nom_Districte  Total_Activitat_Per_Barri  
0              ALTRES    Sant Martí                         20  
1          LOCAL BUIT    Sant Martí                         42  
2             SERVEIS    Sant Martí                        185  
3          LOCAL BUIT    Sant Martí                         42  
4          LOCAL BUIT    Sant Martí                         42

In [34]:
# pivot table para que las actividades economicas se coniertan en columnas y reducir un poco la dimensionalidad
df_cens_pivot = pd.pivot_table(df_cens, values='Total_Activitat_Per_Barri', index=['Codi_Districte','Nom_Districte','Codi_Barri','Nom_Barri'], columns='Activitat_Economica', fill_value=0)

In [35]:
df_cens_pivot.reset_index(inplace=True)

In [36]:
# cambio de nombre de columnas para estar alineados con el equipo de Back
df_cens_pivot.rename(columns = {
    'Codi_Districte':'codiDistricte',
    'Nom_Districte' : 'nomDistricte',
    'Nom_Barri':'nomBarri',
    'Codi_Barri':'codiBarri',
    'Nom_Barri':'nomBarri',
    'ALTRES':'nAltres',
    'AUTOMOCIÓ':'nAutomocio',
    'EQUIPAMENT PERSONAL':'nEquipamentPersonal',
    'LOCAL BUIT' : 'nLocalBuit',
    'OCI I CULTURA': 'nOciCultura',
    'PARAMENT DE LA LLAR':'nParamentLlar',
    'QUOTIDIÀ ALIMENTARI' : 'nQuotidiaAlimentari',
    'QUOTIDIÀ NO ALIMENTARI' : 'nQuotidiaNoAlimentari',
    'SERVEIS':'nServeis'
    
},inplace=True

)

In [37]:
# generamos una nueva columna con el año del dataset para más tarde añadir otro años
df_cens_pivot['Anio'] = '2014'

In [38]:
df_cens_pivot.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73 entries, 0 to 72
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   codiDistricte          73 non-null     object
 1   nomDistricte           73 non-null     object
 2   codiBarri              73 non-null     object
 3   nomBarri               73 non-null     object
 4   nAltres                73 non-null     int64 
 5   nAutomocio             73 non-null     int64 
 6   nEquipamentPersonal    73 non-null     int64 
 7   nLocalBuit             73 non-null     int64 
 8   nOciCultura            73 non-null     int64 
 9   nParamentLlar          73 non-null     int64 
 10  nQuotidiaAlimentari    73 non-null     int64 
 11  nQuotidiaNoAlimentari  73 non-null     int64 
 12  nServeis               73 non-null     int64 
 13  Anio                   73 non-null     object
dtypes: int64(9), object(5)
memory usage: 8.1+ KB


In [39]:
# Convertimos el año en formato fecha
df_cens_pivot['Anio'] = pd.to_datetime(df_cens_pivot['Anio'],format = "%Y")

In [40]:
df_cens_pivot

Activitat_Economica codiDistricte  nomDistricte codiBarri  \
0                              01  Ciutat Vella        01   
1                              01  Ciutat Vella        02   
2                              01  Ciutat Vella        03   
3                              01  Ciutat Vella        04   
4                              02      Eixample        05   
..                            ...           ...       ...   
68                             10    Sant Martí        69   
69                             10    Sant Martí        70   
70                             10    Sant Martí        71   
71                             10    Sant Martí        72   
72                             10    Sant Martí        73   

Activitat_Economica                                      nomBarri  nAltres  \
0                                                        el Raval      381   
1                                                  el Barri Gòtic      315   
2                                                  la Barceloneta       99   
3                           Sant Pere, Santa Caterina i la Ribera      259   
4                                                   el Fort Pienc       81   
..                                                            ...      ...   
68                   Diagonal Mar i el Front Marítim del Poblenou       20   
69                                          el Besòs i el Maresme       90   
70                                        Provençals del Poblenou      166   
71                                       Sant Martí de Provençals       80   
72                                            la Verneda i la Pau       75   

Activitat_Economica  nAutomocio  nEquipamentPersonal  nLocalBuit  nOciCultura  \
0                             1                  225         245           88   
1                             0                  478         199           78   
2                             0                   26         108           24   
3                             1                  363         155           53   
4                            12                   40         199           47   
..                          ...                  ...         ...          ...   
68                            0                    8          42           17   
69                            1                   16          59           12   
70                            5                   11         122           26   
71                            4                   40         120           23   
72                            2                   34         140           26   

Activitat_Economica  nParamentLlar  nQuotidiaAlimentari  \
0                              137                  296   
1                               76                  161   
2                               13                   89   
3                               98                  175   
4                               55                  113   
..                             ...                  ...   
68                               3                   19   
69                              22                   71   
70                              16                   40   
71                              34                   84   
72                              43                  116   

Activitat_Economica  nQuotidiaNoAlimentari  nServeis       Anio  
0                                       47      1120 2014-01-01  
1                                       37       786 2014-01-01  
2                                       16       370 2014-01-01  
3                                       34       688 2014-01-01  
4                                       33       556 2014-01-01  
..                                     ...       ...        ...  
68                                      11       185 2014-01-01  
69                                      18       204 2014-01-01  
70                                      19       314 2

# CensComercialBCN 2016

## Load data

In [41]:
# Import data from API
df_cens_2016 = import_data('18cbddda-ad3f-41b8-8229-064f0faa930b').drop(columns = '_id')
df_cens_2016.head()

NUM_POLICI N_EIX SN_CCOMERC Codi_Districte ID_CCOMERC N_PRINCIP  \
0          5                1              1      CC003     Actiu   
1          5                1              1      CC003     Actiu   
2          5                1              1      CC003     Actiu   
3          5                1              1      CC003     Actiu   
4          5                1              1      CC003     Actiu   

                  DATA    LATITUD SN_GALERIA ID_BCN  ...  LONGITUD  \
0  2016-06-07T00:00:00  41.375377          0     49  ...  2.182944   
1  2016-06-07T00:00:00   41.37539          0     50  ...  2.183092   
2  2016-06-07T00:00:00  41.375404          0     51  ...  2.183217   
3  2016-06-07T00:00:00   41.37542          0     52  ...  2.183354   
4  2016-06-07T00:00:00  41.375425          0     53  ...  2.183517   

           N_SECTOR         REF_CAD  \
0  Comerç al detall  1813205DF3811D   
1           Serveis  1813205DF3811D   
2  Comerç al detall  1813205DF3811D   
3  Comerç al detall  1813205DF3811D   
4           Serveis  1813205DF3811D   

                                           N_GRUPACT       Nom_Barri  \
0                                Equipament personal  el Barri Gòtic   
1  Restaurants, bars i hotels (Inclòs hostals, pe...  el Barri Gòtic   
2                                Equipament personal  el Barri Gòtic   
3                                Equipament personal  el Barri Gòtic   
4  Restaurants, bars i hotels (Inclòs hostals, pe...  el Barri Gòtic   

       N_CARRER N_GALERIA   N_CCOMERC ID_SECTOR ID_GRUPACT  
0  MOLL ESPANYA            Maremàgnum         1          4  
1  MOLL ESPANYA            Maremàgnum         2         14  
2  MOLL ESPANYA            Maremàgnum         1          4  
3  MOLL ESPANYA            Maremàgnum         1          4  
4  MOLL ESPANYA            Maremàgnum         2         14  

[5 rows x 34 columns]

## EDA y preprocesamiento

In [42]:
report_df(df_cens_2016)

El tamaño del dataset es 78033 filas y 34 columnas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78033 entries, 0 to 78032
Data columns (total 34 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   NUM_POLICI      78017 non-null  object
 1   N_EIX           78033 non-null  object
 2   SN_CCOMERC      78033 non-null  object
 3   Codi_Districte  78033 non-null  object
 4   ID_CCOMERC      78033 non-null  object
 5   N_PRINCIP       78033 non-null  object
 6   DATA            78033 non-null  object
 7   LATITUD         78033 non-null  object
 8   SN_GALERIA      78033 non-null  object
 9   ID_BCN          78033 non-null  object
 10  N_MERCAT        78033 non-null  object
 11  N_LOCAL         78033 non-null  object
 12  ID_PRINCIP      78033 non-null  object
 13  SN_CARRER       78033 non-null  object
 14  SEC_CENS        78033 non-null  object
 15  SN_EIX          78033 non-null  object
 16  Y_UTM_ETRS      78033 non-null  object
 17

In [43]:
# Eliminar columnas innecesarias
df_cens_2016 =  df_cens_2016.drop(['NUM_POLICI', 'N_EIX', 'SN_CCOMERC', 'ID_CCOMERC',
                       'N_PRINCIP', 'DATA', 'LATITUD', 'SN_GALERIA', 'ID_BCN', 'N_MERCAT',
                       'N_LOCAL', 'ID_PRINCIP', 'SN_CARRER', 'SEC_CENS', 'SN_EIX',
                       'Y_UTM_ETRS', 'X_UTM_ETRS', 'ID_ACT', 'ID_MERCAT', 
                       'SN_MERCAT', 'LONGITUD', 'REF_CAD',
                       'N_CARRER', 'N_GALERIA', 'N_CCOMERC',
                       'ID_SECTOR', 'ID_GRUPACT'], axis=1)

In [44]:
# Cambiamos el nombre de la columna para que tenga más coherencia con los demás datos con los que trabajamosç
df_cens_2016.rename(columns={'N_DISTRI': 'Nom_Districte'}, inplace=True)

In [45]:
df_cens_2016.describe(exclude=np.number).T

count unique                top   freq
Codi_Districte  78033     10                  2  16179
Nom_Districte   78033     10           Eixample  16179
Codi_Barri      78033     73                 31   4581
N_ACT           78033     62                  -  17768
N_SECTOR        78033      8            Serveis  31773
N_GRUPACT       78033     16             Altres  20867
Nom_Barri       78033     75  la Vila de Gràcia   4581

Después de analizar los valores de las columnas N_DISTRI, N_SECTOR, N_GRUPACT (que contienen información sobre los tipos de actividad economica) nos decantamos por N_GRUPACT.

In [46]:
print("Número de valores únicos: " + str(len(df_cens_2016['N_ACT'].unique())))
df_cens_2016['N_ACT'].value_counts()

Número de valores únicos: 62


-                                 17768
Serveis de menjar i begudes        9680
Vestir                             3296
Resta alimentació                  3266
Altres                             2906
                                  ...  
Música                               84
Combustibles i carburants            75
Fabricació tèxtil                    75
Maquinària                           60
Grans magatzems i hipermercats       30
Name: N_ACT, Length: 62, dtype: int64

In [47]:
print("Número de valores únicos: " + str(len(df_cens_2016['N_SECTOR'].unique())))
df_cens_2016['N_SECTOR'].value_counts()

Número de valores únicos: 8


Serveis                            31773
Comerç al detall                   22747
Sense informació                   13023
Altres                              5745
Locals buits en lloguer             2208
Locals buits en venda i lloguer     1052
Locals buits en venda                787
En reforma                           698
Name: N_SECTOR, dtype: int64

In [48]:
print("Número de valores únicos: " + str(len(df_cens_2016['N_GRUPACT'].unique())))
df_cens_2016['N_GRUPACT'].value_counts()

Número de valores únicos: 16


Altres                                                            20867
-                                                                 17768
Restaurants, bars i hotels (Inclòs hostals, pensions i fondes)    10339
Quotidià alimentari                                                8536
Equipament personal                                                5175
Quotidià no alimentari                                             2181
Parament de la llar                                                2101
Ensenyament                                                        1907
Oci i cultura                                                      1801
Sanitat i assistència                                              1698
Reparacions (Electrodomèstics i automòbils)                        1511
Finances i assegurances                                            1511
Equipaments culturals i recreatius                                 1236
Activitats immobiliàries                                        

In [49]:
df_cens_2016 =  df_cens_2016.drop(['N_ACT', 'N_SECTOR'], axis=1)
df_cens_2016.head(3)

Codi_Districte Nom_Districte Codi_Barri  \
0              1  Ciutat Vella          2   
1              1  Ciutat Vella          2   
2              1  Ciutat Vella          2   

                                           N_GRUPACT       Nom_Barri  
0                                Equipament personal  el Barri Gòtic  
1  Restaurants, bars i hotels (Inclòs hostals, pe...  el Barri Gòtic  
2                                Equipament personal  el Barri Gòtic

Como vemos, N_GRUPACT no tiene exactamente los mismos valores que la clasificación usada en los datos de 2014, por lo tanto cambiamos los valores. Como hemos podido comprobar inspecionando manualmente el anterior dataset, todos los valores nuevos estan classificados como "serveis" en el dataset de 2014. Por lo tanto, los sustituimos.

Por otro lado, también decidimos cambiar los valores faltantes (representados como '-'), por 'Altres'. Hemos tomado esta decisión para evitar perder datos, ya que estamos hablando de 17768 filas, pero evidentemente otra opción hubiera sido eliminarlos.

In [50]:
df_cens_2016['N_GRUPACT'] = df_cens_2016['N_GRUPACT'].replace({'Restaurants, bars i hotels (Inclòs hostals, pensions i fondes)': 'SERVEIS',
                                                    'Ensenyament': 'SERVEIS',
                                                    'Reparacions (Electrodomèstics i automòbils)': 'SERVEIS',
                                                    'Sanitat i assistència': 'SERVEIS',
                                                    'Finances i assegurances': 'SERVEIS',
                                                    'Equipaments culturals i recreatius': 'SERVEIS',
                                                    'Manteniment, neteja i producció': 'SERVEIS',
                                                    'Activitats immobiliàries': 'SERVEIS',
                                                    '-': 'Altres'})

df_cens_2016['N_GRUPACT'].value_counts()

Altres                    38635
SERVEIS                   19331
Quotidià alimentari        8536
Equipament personal        5175
Quotidià no alimentari     2181
Parament de la llar        2101
Oci i cultura              1801
Automoció                   273
Name: N_GRUPACT, dtype: int64

Hemos detectado que hay algun error en los barrios ya que salen 75. Después de investigarlo hemos llegado a la conclusión de hay barrios repetidos pero con diferencias en mayúsculas y minúsculas, así que pasamos todo a minúsculas.

In [51]:
df_cens_2016.describe(exclude=np.number).T

count unique                top   freq
Codi_Districte  78033     10                  2  16179
Nom_Districte   78033     10           Eixample  16179
Codi_Barri      78033     73                 31   4581
N_GRUPACT       78033      8             Altres  38635
Nom_Barri       78033     75  la Vila de Gràcia   4581

In [52]:
df_cens_2016['Nom_Barri'] = df_cens_2016['Nom_Barri'].str.lower()
len(df_cens_2016['Nom_Barri'].value_counts())

73

In [53]:
df_cens_2016

Codi_Districte        Nom_Districte Codi_Barri            N_GRUPACT  \
0                  1         Ciutat Vella          2  Equipament personal   
1                  1         Ciutat Vella          2              SERVEIS   
2                  1         Ciutat Vella          2  Equipament personal   
3                  1         Ciutat Vella          2  Equipament personal   
4                  1         Ciutat Vella          2              SERVEIS   
...              ...                  ...        ...                  ...   
78028              5  Sarrià-Sant Gervasi         26  Equipament personal   
78029              5  Sarrià-Sant Gervasi         26  Parament de la llar   
78030              5  Sarrià-Sant Gervasi         26               Altres   
78031              5  Sarrià-Sant Gervasi         26              SERVEIS   
78032              4            Les Corts         19               Altres   

                   Nom_Barri  
0             el barri gòtic  
1             el barri gòtic  
2             el barri gòtic  
3             el barri gòtic  
4             el barri gòtic  
...                      ...  
78028  sant gervasi  galvany  
78029  sant gervasi  galvany  
78030  sant gervasi  galvany  
78031  sant gervasi  galvany  
78032              les corts  

[78033 rows x 5 columns]

In [54]:
# Cambio nombre columna por un nombre más adecuado al contenido
df_cens_2016.rename(columns = {'N_GRUPACT':'Activitat_Economica'},inplace=True)

## Cálculo de actividad por barrio

In [55]:
# Groupby para calcular el total de actividades que hay según el barrio y el tipo de actividad economica
Activitat_Barris_2016 = df_cens_2016.groupby(['Codi_Barri','Activitat_Economica']).size().to_frame(name = 'Total_Activitat_Per_Barri').reset_index()
Activitat_Barris_2016

Codi_Barri     Activitat_Economica  Total_Activitat_Per_Barri
0            1                  Altres                       1316
1            1               Automoció                          1
2            1     Equipament personal                        208
3            1           Oci i cultura                        124
4            1     Parament de la llar                         88
..         ...                     ...                        ...
530          9           Oci i cultura                         77
531          9     Parament de la llar                         74
532          9     Quotidià alimentari                        236
533          9  Quotidià no alimentari                         70
534          9                 SERVEIS                        741

[535 rows x 3 columns]

In [56]:
# merge del dataframe original con el dataframe generado con el groupby anterior. Comprobamos que no se haga un cartesiano
print(df_cens_2016.shape)

df_cens_2016 = pd.merge(df_cens_2016,Activitat_Barris_2016, on = ['Codi_Barri','Activitat_Economica'], how = 'left')

print(df_cens_2016.shape)

(78033, 5)
(78033, 6)


In [57]:
# pivot table para que las actividades economicas se coniertan en columnas y reducir un poco la dimensionalidad
df_cens_pivot_2016 = pd.pivot_table(df_cens_2016, values='Total_Activitat_Per_Barri', index=['Codi_Districte','Nom_Districte','Codi_Barri','Nom_Barri'], columns='Activitat_Economica', fill_value=0)
df_cens_pivot_2016.reset_index(inplace=True)

In [58]:
df_cens_pivot_2016

Activitat_Economica Codi_Districte Nom_Districte Codi_Barri  \
0                                1  Ciutat Vella          1   
1                                1  Ciutat Vella          2   
2                                1  Ciutat Vella          3   
3                                1  Ciutat Vella          4   
4                               10    Sant Martí         64   
..                             ...           ...        ...   
68                               9   Sant Andreu         59   
69                               9   Sant Andreu         60   
70                               9   Sant Andreu         61   
71                               9   Sant Andreu         62   
72                               9   Sant Andreu         63   

Activitat_Economica                              Nom_Barri  Altres  Automoció  \
0                                                 el raval    1316          1   
1                                           el barri gòtic     817          0   
2                                           la barceloneta     256          0   
3                    sant pere, santa caterina i la ribera     812          1   
4                               el camp de l'arpa del clot     946          9   
..                                                     ...     ...        ...   
68                                           el bon pastor     474         20   
69                                             sant andreu    1180          4   
70                                              la sagrera     465          3   
71                                el congrés i els indians     364          1   
72                                                   navas     357          5   

Activitat_Economica  Equipament personal  Oci i cultura  Parament de la llar  \
0                                    208            124                   88   
1                                    484             52                   49   
2                                     19             12                    7   
3                                    290             35                   66   
4                                     85             36                   55   
..                                   ...            ...                  ...   
68                                   104             16                   23   
69                                   124             52                   67   
70                                    34             16                   21   
71                                    22              9                   17   
72                                    30             16                   39   

Activitat_Economica  Quotidià alimentari  Quotidià no alimentari  SERVEIS  
0                                    574                      58      660  
1                                    151                      43      560  
2                                    110                      16      248  
3                                    254                      40      464  
4                                    175                      52      376  
..                                   ...                     ...      ...  
68                                    52                      32      158  
69                                   237                      69      533  
70                                   131                      30      213  
71                                    52                      20      180  
72                                    70                      19      152  

[73 rows x 12 columns]

In [59]:
# cambio de nombre de columnas para estar alineados con el equipo de Back
df_cens_pivot_2016.rename(columns = {
    'Codi_Districte':'codiDistricte',
    'Nom_Districte' : 'nomDistricte',
    'Nom_Barri':'nomBarri',
    'Codi_Barri':'codiBarri',
    'Nom_Barri':'nomBarri',
    'Altres':'nAltres',
    'Automoció':'nAutomocio',
    'Equipament personal':'nEquipamentPersonal',
    'Oci i cultura': 'nOciCultura',
    'Parament de la llar':'nParamentLlar',
    'Quotidià alimentari' : 'nQuotidiaAlimentari',
    'Quotidià no alimentari' : 'nQuotidiaNoAlimentari',
    'SERVEIS':'nServeis'
    
},inplace=True

)

In [60]:
# generamos una nueva columna con el año del dataset
df_cens_pivot_2016['Anio'] = '2016'

In [61]:
df_cens_pivot_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73 entries, 0 to 72
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   codiDistricte          73 non-null     object
 1   nomDistricte           73 non-null     object
 2   codiBarri              73 non-null     object
 3   nomBarri               73 non-null     object
 4   nAltres                73 non-null     int64 
 5   nAutomocio             73 non-null     int64 
 6   nEquipamentPersonal    73 non-null     int64 
 7   nOciCultura            73 non-null     int64 
 8   nParamentLlar          73 non-null     int64 
 9   nQuotidiaAlimentari    73 non-null     int64 
 10  nQuotidiaNoAlimentari  73 non-null     int64 
 11  nServeis               73 non-null     int64 
 12  Anio                   73 non-null     object
dtypes: int64(8), object(5)
memory usage: 7.5+ KB


In [62]:
# Convertimos el año en formato fecha
df_cens_pivot_2016['Anio'] = pd.to_datetime(df_cens_pivot_2016['Anio'],format = "%Y")

In [63]:
df_cens_pivot_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73 entries, 0 to 72
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   codiDistricte          73 non-null     object        
 1   nomDistricte           73 non-null     object        
 2   codiBarri              73 non-null     object        
 3   nomBarri               73 non-null     object        
 4   nAltres                73 non-null     int64         
 5   nAutomocio             73 non-null     int64         
 6   nEquipamentPersonal    73 non-null     int64         
 7   nOciCultura            73 non-null     int64         
 8   nParamentLlar          73 non-null     int64         
 9   nQuotidiaAlimentari    73 non-null     int64         
 10  nQuotidiaNoAlimentari  73 non-null     int64         
 11  nServeis               73 non-null     int64         
 12  Anio                   73 non-null     datetime64[ns]
dtypes: date

In [64]:
df_cens_pivot_2016

Activitat_Economica codiDistricte  nomDistricte codiBarri  \
0                               1  Ciutat Vella         1   
1                               1  Ciutat Vella         2   
2                               1  Ciutat Vella         3   
3                               1  Ciutat Vella         4   
4                              10    Sant Martí        64   
..                            ...           ...       ...   
68                              9   Sant Andreu        59   
69                              9   Sant Andreu        60   
70                              9   Sant Andreu        61   
71                              9   Sant Andreu        62   
72                              9   Sant Andreu        63   

Activitat_Economica                               nomBarri  nAltres  \
0                                                 el raval     1316   
1                                           el barri gòtic      817   
2                                           la barceloneta      256   
3                    sant pere, santa caterina i la ribera      812   
4                               el camp de l'arpa del clot      946   
..                                                     ...      ...   
68                                           el bon pastor      474   
69                                             sant andreu     1180   
70                                              la sagrera      465   
71                                el congrés i els indians      364   
72                                                   navas      357   

Activitat_Economica  nAutomocio  nEquipamentPersonal  nOciCultura  \
0                             1                  208          124   
1                             0                  484           52   
2                             0                   19           12   
3                             1                  290           35   
4                             9                   85           36   
..                          ...                  ...          ...   
68                           20                  104           16   
69                            4                  124           52   
70                            3                   34           16   
71                            1                   22            9   
72                            5                   30           16   

Activitat_Economica  nParamentLlar  nQuotidiaAlimentari  \
0                               88                  574   
1                               49                  151   
2                                7                  110   
3                               66                  254   
4                               55                  175   
..                             ...                  ...   
68                              23                   52   
69                              67                  237   
70                              21                  131   
71                              17                   52   
72                              39                   70   

Activitat_Economica  nQuotidiaNoAlimentari  nServeis       Anio  
0                                       58       660 2016-01-01  
1                                       43       560 2016-01-01  
2                                       16       248 2016-01-01  
3                                       40       464 2016-01-01  
4                                       52       376 2016-01-01  
..                                     ...       ...        ...  
68                                      32       158 2016-01-01  
69                                      69       533 2016-01-01  
70                                      30       213 2016-01-01  
71                                      20       180 2016-01-01  
72                                      19       152 2016-01-01  

[73 rows x 13 columns]

## Concatenar datasets

Concatenamos con el dataset anterior para tener la información sobre los dos años.

In [65]:
# Concatenamos con el anterior dataset
df_cens_concat = pd.concat([df_cens_pivot, df_cens_pivot_2016], ignore_index=True)
df_cens_concat

codiDistricte  nomDistricte codiBarri  \
0              01  Ciutat Vella        01   
1              01  Ciutat Vella        02   
2              01  Ciutat Vella        03   
3              01  Ciutat Vella        04   
4              02      Eixample        05   
..            ...           ...       ...   
141             9   Sant Andreu        59   
142             9   Sant Andreu        60   
143             9   Sant Andreu        61   
144             9   Sant Andreu        62   
145             9   Sant Andreu        63   

                                  nomBarri  nAltres  nAutomocio  \
0                                 el Raval      381           1   
1                           el Barri Gòtic      315           0   
2                           la Barceloneta       99           0   
3    Sant Pere, Santa Caterina i la Ribera      259           1   
4                            el Fort Pienc       81          12   
..                                     ...      ...         ...   
141                          el bon pastor      474          20   
142                            sant andreu     1180           4   
143                             la sagrera      465           3   
144               el congrés i els indians      364           1   
145                                  navas      357           5   

     nEquipamentPersonal  nLocalBuit  nOciCultura  nParamentLlar  \
0                    225       245.0           88            137   
1                    478       199.0           78             76   
2                     26       108.0           24             13   
3                    363       155.0           53             98   
4                     40       199.0           47             55   
..                   ...         ...          ...            ...   
141                  104         NaN           16             23   
142                  124         NaN           52             67   
143                   34         NaN           16             21   
144                   22         NaN            9             17   
145                   30         NaN           16             39   

     nQuotidiaAlimentari  nQuotidiaNoAlimentari  nServeis       Anio  
0                    296                     47      1120 2014-01-01  
1                    161                     37       786 2014-01-01  
2                     89                     16       370 2014-01-01  
3                    175                     34       688 2014-01-01  
4                    113                     33       556 2014-01-01  
..                   ...                    ...       ...        ...  
141                   52                     32       158 2016-01-01  
142                  237                     69       533 2016-01-01  
143                  131                     30       213 2016-01-01  
144                   52                     20       180 2016-01-01  
145                   70                     19       152 2016-01-01  

[146 rows x 14 columns]

Como vemos, en este caso faltaria rellenar los datos de 'nLocalBuit' para el segundo dataset ya que no estaban presentes en la columna escogida para la actividad económica. Sería el siguiente paso a hacer. También se deberia tener en cuenta la diferencia en los nombres de les barrios.

# CensComercialBCN 2019

In [70]:
df_cens_2019 = import_data('18cbddda-ad3f-41b8-8229-064f0faa930b').drop(columns = '_id')
df_cens_2019.head()

NUM_POLICI N_EIX SN_CCOMERC Codi_Districte ID_CCOMERC N_PRINCIP  \
0          5                1              1      CC003     Actiu   
1          5                1              1      CC003     Actiu   
2          5                1              1      CC003     Actiu   
3          5                1              1      CC003     Actiu   
4          5                1              1      CC003     Actiu   

                  DATA    LATITUD SN_GALERIA ID_BCN  ...  LONGITUD  \
0  2016-06-07T00:00:00  41.375377          0     49  ...  2.182944   
1  2016-06-07T00:00:00   41.37539          0     50  ...  2.183092   
2  2016-06-07T00:00:00  41.375404          0     51  ...  2.183217   
3  2016-06-07T00:00:00   41.37542          0     52  ...  2.183354   
4  2016-06-07T00:00:00  41.375425          0     53  ...  2.183517   

           N_SECTOR         REF_CAD  \
0  Comerç al detall  1813205DF3811D   
1           Serveis  1813205DF3811D   
2  Comerç al detall  1813205DF3811D   
3  Comerç al detall  1813205DF3811D   
4           Serveis  1813205DF3811D   

                                           N_GRUPACT       Nom_Barri  \
0                                Equipament personal  el Barri Gòtic   
1  Restaurants, bars i hotels (Inclòs hostals, pe...  el Barri Gòtic   
2                                Equipament personal  el Barri Gòtic   
3                                Equipament personal  el Barri Gòtic   
4  Restaurants, bars i hotels (Inclòs hostals, pe...  el Barri Gòtic   

       N_CARRER N_GALERIA   N_CCOMERC ID_SECTOR ID_GRUPACT  
0  MOLL ESPANYA            Maremàgnum         1          4  
1  MOLL ESPANYA            Maremàgnum         2         14  
2  MOLL ESPANYA            Maremàgnum         1          4  
3  MOLL ESPANYA            Maremàgnum         1          4  
4  MOLL ESPANYA            Maremàgnum         2         14  

[5 rows x 34 columns]

Para los datos del 2019 seguiriamos el mismo procedimiento que como hasta ahora, seleccionando las columnas de barrio, distrito y actividad economica correspondientes. Este proceso quedaría pendiente como trabajo futuro en el caso de disponer de más tiempo.

# TO JSON

In [67]:
# convertimos nuestro dataset a formato json para pasar al equipo de Back
df_censo_json = df_cens_concat.to_json(orient='records', lines=True)

In [68]:
file_path = "..\outputs\df_censo_json.json"

with open(file_path, 'w') as file:
    file.write(df_censo_json)